### ISSUE TITLE
Memory leak when FormControlName created/destroyed few times

### ISSUE THREAD

**Issue by IliaVolk** [angular/angular#20007]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

1. FormControlName, objects provided by NG_VALIDATORS (may be some other) can not be collected by gc until there are refs for FormGroup.
This is caused by collecting closures created in https://github.com/angular/angular/blob/4.4.6/packages/forms/src/directives/shared.ts#L33  in https://github.com/angular/angular/blob/4.4.6/packages/forms/src/model.ts#L744 .
2. All provided validators (via NG_VALIDATORS token) are called as many times as form is recreated.

## Expected behavior
1. After destroying FormControlName it should be collected by gc even if there are refs for FormGroup.
2. Provided validators should be removed from FormControl after destroying FormControlName.

## Minimal reproduction of the problem with instructions
https://plnkr.co/edit/eCtf2naoxLe8bWTJ8IkV?p=preview (used example from material demo as starting point)

#### Steps to reproduce
1. show/hide form clicking checkbox.
2. displayed below number of listeners is increasing.
3. there are more and more messages 'Validate called' appears in console when editing input (In my project I had issue when validators set outdated errors, had to use workaround).

There are 3 inputs to show that they are not synchronized (Is there a significant reason to store all closures, not only last?).

## What is the motivation/use case for changing the behavior?

1. Allow to have permanent FormControl/FormGroup without memory leaks.
2. Allow to dynamically create FormControlName with custom control, which knows how to validate itself (providing NG_VALIDATORS), without calling validators of destroyed controls.
## Environment

<pre><code>
Angular version: 4.4.6

Browser:
- [x] Chrome (desktop) version 62.0.3202.75
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
Others:

</code></pre>

**Comment by aalbericio** [comment]
We've hit this bug.

We have our form's creation/destruction affected by an ngSwitch. If we use formControlName in our inputs, hiding and showing (creating and destroying) the form causes Subjects and SubjectSubscriptions to grow and never dispose.

It's been a hard time to debug this and memory leaks are always important.

I hope this gets fixed as soon as possible.

Thanks

**Comment by aalbericio** [comment]
We can provide a stackblitz project if needed.

**Comment by vijosoco** [comment]
We have run into this leak and it a pretty big deal - any ETA on when this might be fixed and available.

Anybody know of a workaround ?

**Comment by vkrepkiy** [comment]
Just ran into this bug. I have custom input with the NG_VALIDATORS provided. If user enters invalid data and navigates to the other tab of the same form, causing this custom input to be destroyed, he is unable to submit the form ever again without navigating out (force form to be destroyed). I think this bug is really severe.

**Comment by gdnext** [comment]
I recently ran into this problem with a rather large project and its causing a major performance issue. Any updates here would put my mind at ease.

**Comment by thomas-parrish** [comment]
Why has such a critical issue been open for a year without being addressed? This is not just a memory leak, this makes it impossible to submit any form on our site after a user enters invalid input once. 

Has anyone discovered a workaround?

**Comment by IliaVolk** [comment]
Workaround - do not use html attribute validators with formControlName and recreate FormGroup instance when big memory leak is possible

**Comment by pengx17** [comment]
The plunker example is dead. Here is a minimized stackblitz version:
https://stackblitz.com/edit/angular-jr5cds?file=src%2Fapp%2Fcontrol.component.html

The workaround we can come up with for now is to use hidden attribute instead of ngIf for the snippet of form to enforce stabilized structure of the reactive formgroup.

**Comment by longgt** [comment]
Workaround for me now is patch FormControlName.ngOnDestroy manually.

~~~
export function patchNgFormControlName () {
    return function () {
        const _prototype: any = FormControlName.prototype as any;
        const PATCHED_NAME: string = '__form_control_name_patched';
        
        if (!_prototype[PATCHED_NAME]) {
            const original = _prototype.ngOnDestroy;
            _prototype.ngOnDestroy = function () {
                original.apply(this, arguments);
                if (this.control && typeof this.control._clearChangeFns === 'function') {
                    this.control._clearChangeFns();
                }
            };
            _prototype[PATCHED_NAME] = true;
        }
    };
}
~~~
and add the above function as APP_INITIALIZER in app.module

**Comment by maxime1992** [comment]
@longgt as explained by @devoto13 [here](https://github.com/angular/angular/pull/22503#issuecomment-452021464) that may cause unexpected behavior. After digging a bit into the code I think I understood why:
- every FormControlDirective will create an instance of a FormControl
- if you create multiple FormControlDirective that have the same formControl or FormControlName they will refer to the same FormControl instance
- within your example, destroying one of those directive attached to a component would trigger the `_clearChangeFns` that will clear everything for all the others too

I agree that it might not be super common to have multiple FormControlDirective pointing to the same FormControl but still, just a warning :) 

A proper fix for that is doable but not trivial either...

**Comment by jandeu** [comment]
Unfortunately this is not just a memory leak. It can also corrupt form data and it has in our application. Let me show how:
 - Let's have a component with two instances of FormControl  stored in fields A and B and one input in component's view.
 - Now if we bind input with form control A and then change it to B everything works fine. Unless we update the value of control A. If we do, the input element will be updated with value of A but any change made to it by user will be saved in value of control B.
Here is a [StackBlitz](https://stackblitz.com/edit/angular-6aqeqc)

**Comment by MaciejWierzchowski** [comment]
Is there any other workaround, i.e. which doens't require to monkey patch angular?

**Comment by maxime1992** [comment]
This issue has been around for nearly 3 years now.
_(I usually don't like to start a message this way unless I tried something to fix the issue myself... [Which I did here](https://github.com/angular/angular/issues/20007#issuecomment-452635900) :grimacing: :angel:! and failed miserably as it seem quite complex to get to the bottom of it... But bear with me)_

I do know that Ivy was necessary in order to solve a lot of issues in Angular ecosystem and that it ended up being really time and resource consumming for the Angular team to ensure that it was working well, that it was fast, that it was retro compatible, that ngcc could pre compile the libraries, etc. I trully appreciate the effort here :clap: :clap: :clap:.

On the other hand, I think that like many other devs over the last few years, I've felt an important frustration with Angular.
Not because I wanted more brilliant features.
Not because Ivy wasn't there quite yet and I couldn't lazy load a component without using the router.
But rather because we've had **annoying, recognized and long lived bugs, sometimes with lot of comments from the community to help debug or even PR made by the community to fix them (:warning: !) pending for months or even years without any action or comment** to fix them and bring them back into the main branch. So yes, pretty much since Ivy development started it's been quite a frustrating experience.

_(I needed to get that off my chest at some point and did it into this issue, sorry :man_shrugging:)_

That said, I've always loved Angular community and always hoped that with Ivy going stable _(I know the dev is not finished 100% on that though)_ it'd bring slowly but surely more focus on bug fixes, then new features :pray:. When I saw this tweet: https://twitter.com/angular/status/1265342327351939073 I was super happy as it felt like this was the beginning of it :raised_hands:. Time will tell but still, I've seen a lot of efforts put into long lived and major issues (+ features!) so keep it up :muscle: :fire: :pray:.
 
Now, with the recent efforts from the Angular team to apparently tackle such important, annoying, long lived and forgotten issues I have hope that some light will be brought upon this one. **This memory leak is making our app miserable :disappointed:**. Let me share why.

We have hundreds of forms into our app. We need to provide a lot of data from the "real world" to our app in order to run a simulation. **Those forms are huge**. In terms of data they contain and have to expose, types _(input, radio buttons, checkboxes, select, ...)_. The UI components are also quite huge. For example we can have SVG rendered next to the form to represent some complex shapes. This way the user is able to visualise if the data he's entering make sense.

In order to manage all those data, we did try a lot of different solutions to **break down our huge forms into smaller ones**. We even have some "sub forms" that we can reuse in other forms :fire:. As this was critical for us and after spending an important amount of time to find a solution that was a nice fit for us, we decided to open source a library we built ourselves:
**ngx-sub-form** https://github.com/cloudnc/ngx-sub-form

This library internally uses [ControlValueAccessor](https://angular.io/api/forms/ControlValueAccessor) to build the sub forms and while **I'm not 100% sure** the issue is coming from there, I believe the odds are quite high. To give you an idea how bad it is for us with all the DOM elements that appear to never be removed, here's a recording of the performance monitor where I:
- Open the app
- Go to a page (let's call it P1) where we've got huge forms (containing lot of sub forms too) 
- Go to a different page (let's call it P2) where we also have huge forms (containing lot of sub forms too) 
- Go to P1
- Go to P2
- Go to P1
- Go to P2

Just by doing that, have a look into the numbers for the DOM nodes and the event listeners :scream: 

![ng-app-memory-leak-3](https://user-images.githubusercontent.com/4950209/91581658-01cb2480-e94f-11ea-90f0-58cbf53ac1a4.gif)

Just by opening 2 different pages, 3 times each, we end up with:
- **+80k detached DOM nodes**
- **+4.9K detached JS event listeners**

**This never goes away and keeps increasing until we reload the page.**

The overall feeling on the app is not pleasant for our users when after 6 clicks the memory footprint is like the following:

![image](https://user-images.githubusercontent.com/4950209/91580759-d7c53280-e94d-11ea-88b7-e8d6b2db0f4b.png)

Now, with all the forms we've got, we may have forgotten something and I may be saying all that for nothing. But after trying to debug this for hours, making sure our observables are closed correctly when the components are destroyed and everything is well cleaned up etc, I can only see this happening on page containing forms (and sub forms using our library). Therefore I'm highly suspicious and think it may not be coming from us. If that's the case, it seems pretty bad and I hope the priority can be bumped so it gets looked at sooner than later.

Now, I know that just asking for something to be fixed is not enough especially when you can't provide a repro.
I can't give away the code of the app I mentioned above BUT we've built a demo app for ngx-sub-form in order to make sure that all the features are working correctly and we run some E2E tests onto that demo etc.

I checked on that demo which you can find here:
- Live demo https://cloudnc.github.io/ngx-sub-form
- Code https://github.com/cloudnc/ngx-sub-form/tree/master/src/app

Note: the forms are much smaller on that demo, contain way less data and have less graphical elements. Therefore the memory leak doesn't seem as bad here but still. 

Info: When clicking on an item on the left, it display into a form (composed of sub forms using `ControlValueAccessor`) the data

![image](https://user-images.githubusercontent.com/4950209/91583014-e8c37300-e950-11ea-9605-7e7aba2e2136.png)

When we switch from one item to another, we get roughly +250 dettached DOM nodes and +100 JS event listeners. So still something that can be used to dig into the issue I assume.

I know everyone love a minimal repro too, so I also made one:

https://stackblitz.com/edit/ng-memory-leak-control-value-accessor?file=src/app/app.component.html

And here's the app live: https://ng-memory-leak-control-value-accessor.stackblitz.io/ 

If you open the chrome dev tool, press `esc` and a new panel will open. In the left menu on that panel, choose to display "Performance monitor" and click on the toggle button multiple times.

![ng-app-memory-leak-4](https://user-images.githubusercontent.com/4950209/91589782-9e46f400-e95a-11ea-8068-49324d5d8a97.gif)

Hope it helps :upside_down_face:

**Comment by rppig42** [comment]
Honestly speaking，developing "fancy" new features may attract a few new users, but setting aside fundamental issues like this one are making more serious users abandoning Angular.

**Comment by xhtml-ru** [comment]
it has been the fourth year since the opening of this issue

**Comment by AndrewKushnir** [comment]
Hi,

Just want to give an update on memory leak issues with ReactiveForms directives.

Over the last few months we landed several updates to internal data structures in Forms package to have an ability to do proper cleanup when Forms-related directives are destroyed. With the upcoming fix in PR #39235 (that relies on the mentioned refactoring), the situation with memory leaks in ReactiveForms directives should be improved. We did verify that there are no memory leaks in common scenarios (and extended test coverage to check the cleanup logic), but we’d like to ask if you could test the changes from PR #39235 in use-cases where you’ve observed the issue and let us know if memory leaks are gone and your app behaves as expected.

Here are few steps to test the changes from PR #39235 with your app:

- make sure the app is upgraded to the latest major version (v11)
- update the `package.json` file in the root folder of the app and use the following value for the ”@angular/forms” package:

```
  “@angular/forms”: “https://875212-24195339-gh.circle-artifacts.com/0/angular/forms-pr39235-5152d0fa00.tgz”,
```
- Run ”npm install” to update the “@angular/forms” locally
- Build your app and test the changes

Note: the mentioned URL is the “@angular/forms” package built by CI with the fixes from PR #39235. This package is **NOT** suitable for production usage and is available only for limited period of time (see more info on build artifacts in [DEVELOPER.md file](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#getting-packages-from-build-artifacts)). The mentioned fix covers only ReactiveForms directives (such as `[formGroup]`, `[formConrol]`, `formControlName`, etc). If a memory leak that you see is related to the template-driven directives such as `ngModel` (and there is no regression with the fix from PR #39235), please check issue tracker for an existing issue or create a new ticket if such issue was not reported previously.

Thank you.

**Comment by maxime1992** [comment]
I'm keen to give it a go @AndrewKushnir :+1:  
**Thanks a lot for all the clean up and bug fixes you've been doing lately :fire:** 

I wanted to try first on the minimal repro I've built here https://stackblitz.com/edit/ng-memory-leak-control-value-accessor-ng11-fix-test but I got an error trying to update with `“@angular/forms”: “https://875212-24195339-gh.circle-artifacts.com/0/angular/forms-pr39235-5152d0fa00.tgz”,` on stackblitz.

So I just downloaded it and ran the project locally... 

| Before upgrading to `“@angular/forms”: “https://875212-24195339-gh.circle-artifacts.com/0/angular/forms-pr39235-5152d0fa00.tgz”,` | After upgrading to `“@angular/forms”: “https://875212-24195339-gh.circle-artifacts.com/0/angular/forms-pr39235-5152d0fa00.tgz”,` |
| --------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------- |
| ![image](https://user-images.githubusercontent.com/4950209/101485677-c67fbf00-395b-11eb-98cf-82ef46931a68.png)                    | ![image](https://user-images.githubusercontent.com/4950209/101486261-ab617f00-395c-11eb-9edf-a1f2c6051c27.png)  |

And I'd say it's looking pretty good :star_struck: :tada: :tada: :tada: :tada: 

Nice one @AndrewKushnir :clap:!

_Note: I don't think this would behave any differently on the app [where I've got a crazy memory leak](https://github.com/angular/angular/issues/20007#issuecomment-682831198) but I haven't been able to test on this one as we're still on angular 10_

**Comment by AndrewKushnir** [comment]
@maxime1992 thanks for trying out the fix and sharing the results!

> Note: I don't think this would behave any differently on the app where I've got a crazy memory leak but I haven't been able to test on this one as we're still on angular 10

I think you could also try the fix with Angular v10. There is nothing specific to v11 in that fix, I just wanted to make sure the `core` and `compiler` packages are relatively up-to-date (so you don't come across potential issues fixed since v10). If you get a chance to test the fix using Angular v10 and your real app, that'd be great!

**Comment by maxime1992** [comment]
@AndrewKushnir I tried in our app to upgrade only `@angular/forms` (and I made sure with the yarn.lock files that it was the one in use) but unfortunately even forcing a garbage collection didn't change anything. 

FYI for the rest of the app we're using angular `10.1.5`. We're using NX, as soon as they publish an upgrade for Angular 11 we'll give it a go and I'll report back :+1:

**Comment by AndrewKushnir** [comment]
@maxime1992 thanks for the update. Would it be possible to create a minimum repro using Angular v10 (and the `@angular/forms` package from PR #39235) based on the use-case that you have in the app? In this case we can try to upgrade it to v11 and perform additional investigation if needed. Note: v11 brings some memory-related fixes (and more fixes has landed in the recent patch version), but if it's purely Forms-related issue, upgrading to v11 would not provide any improvement.

**Comment by AndrewKushnir** [comment]
Just wanted to give a quick update: we've got confirmations that the fix from PR #39235 resolves memory leaks and other problems related to incomplete cleanup in a number of use-cases. We plan to complete testing early Jan'21 and try to land the change in the next minor version (v11.1.0).

@maxime1992, please let us know if you get a chance to perform further investigation using your real app. I'd be very curious to perform further investigation if there is a minimal repro (based on v10 or v11). If you don't have a chance to look into this - no problem, we can always look at other use-cases once PR #39235 lands.

Thank you.

**Comment by paulmartis** [comment]
Hello all,

I can confirm that in my use case, I see my memory leak being solved after I used the following package:
  _“@angular/forms”: “https://875212-24195339-gh.circle-artifacts.com/0/angular/forms-pr39235-5152d0fa00.tgz”,_ (not available anymore)
I am using Angular11. 

So I cannot wait for this to be released soon :) 
Tnx @AndrewKushnir

**Comment by AndrewKushnir** [comment]
Quick update: the fix (PR #39235) was merged and the changes should be available in v11.1.0 release later this month. 
Thanks everyone for helping with investigation and testing 👍 

@maxime1992, please let me know if the memory leak still exists in your app after upgrading to v11.1.0 (once it's released later this month). We'll perform further investigation if needed.
